dist/lib/discovery.js

```diff
@@ -73,24 +75,86 @@ class SKServer {
 exports.SKServer = SKServer;
 
 class Discovery extends _eventemitter.default {
-  constructor(bonjour) {
+  constructor(bonjourOrMdns) {
     let timeout = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 60000;
     super();
-    const props = ['_server', '_registry'].join(',');
+    this.found = [];
 
-    if (!bonjour || typeof bonjour !== 'object' || Object.keys(bonjour).join(',') !== props) {
-      throw new Error('Invalid mDNS provider');
+    if (!bonjourOrMdns || typeof bonjourOrMdns !== 'object') {
+      throw new Error('No mDNS provider given');
     }
 
-    this.found = [];
+    const bonjourProps = ['_server', '_registry'].join(',');
+    const mdnsProps = ['dns_sd', 'Advertisement', 'createAdvertisement', 'Browser'].join(',');
+
+    if (Object.keys(bonjourOrMdns).join(',').startsWith(bonjourProps)) {
+      return this.discoverWithBonjour(bonjourOrMdns, timeout);
+    }
+
+    if (Object.keys(bonjourOrMdns).join(',').startsWith(mdnsProps)) {
+      return this.discoverWithMdns(bonjourOrMdns, timeout);
+    }
+
+    throw new Error('Unrecognized mDNS provider given');
+  }
+
+  discoverWithBonjour(bonjour, timeout) {
     const browser = bonjour.find({
       type: 'signalk-http'
     });
     browser.on('up', ad => {
-      const service = _objectSpread({}, ad.txt, {
+      const service = _objectSpread(_objectSpread({}, ad.txt), {}, {
         name: ad.name || '',
         hostname: ad.host || '',
-        port: parseInt(ad.port, 10)
+        port: parseInt(ad.port, 10),
+        provider: 'bonjour'
+      });
+
+      if (service.hasOwnProperty('roles') && typeof service.roles === 'string' && service.roles.includes(',')) {
+        service.roles = service.roles.split(',').map(role => role.trim().toLowerCase());
+      }
+
+      if (service.hasOwnProperty('roles') && typeof service.roles === 'string' && !service.roles.includes(',')) {
+        service.roles = [service.roles].map(role => role.trim().toLowerCase());
+      }
+
+      let ipv4 = service.hostname;
+
+      if (Array.isArray(ad.addresses)) {
+        ipv4 = ad.addresses.reduce((found, address) => {
+          if (address && typeof address === 'string' && address.includes('.')) {
+            found = address;
+          }
+
+          return found;
+        }, service.hostname);
+      }
+
+      if (ipv4.trim() !== '') {
+        service.hostname = ipv4;
+      }
+
+      const server = new SKServer(service);
+      this.found.push(server);
+      this.emit('found', server);
+    });
+    browser.start();
+    setTimeout(() => {
+      if (this.found.length === 0) {
+        this.emit('timeout');
+      }
+
+      browser.stop();
+    }, timeout);
+  }
+
+  discoverWithMdns(mDNS, timeout) {
+    const browser = mDNS.createBrowser(mDNS.tcp('_signalk-http'));
+    browser.on('serviceUp', ad => {
```

Comment:
In my case I can't get mdns to work with the signalk-js-client, it's probably just a problem on my configuration.
The server announcing with mdns works fine.
With this branch "feature/mdns-backwards-compatibility" the client reuses mdns when called by the server but due to my mdns error the server crashes.

```
events.js:174
      throw er; // Unhandled 'error' event
      ^

Error: getaddrinfo -3008
    at errnoException (/usr/lib/node_modules/signalk-server/node_modules/mdns/lib/resolver_sequence_tasks.js:199:11)
    at getaddrinfo_complete (/usr/lib/node_modules/signalk-server/node_modules/mdns/lib/resolver_sequence_tasks.js:112:10)
    at GetAddrInfoReqWrap.oncomplete (/usr/lib/node_modules/signalk-server/node_modules/mdns/lib/resolver_sequence_tasks.js:120:9)
Emitted 'error' event at:
    at on_resolver_done (/usr/lib/node_modules/signalk-server/node_modules/mdns/lib/browser.js:28:14)
    at next (/usr/lib/node_modules/signalk-server/node_modules/mdns/lib/browser.js:100:7)
    at /usr/lib/node_modules/signalk-server/node_modules/mdns/lib/resolver_sequence_tasks.js:160:11
    at getaddrinfo_complete (/usr/lib/node_modules/signalk-server/node_modules/mdns/lib/resolver_sequence_tasks.js:112:7)
    at GetAddrInfoReqWrap.oncomplete (/usr/lib/node_modules/signalk-server/node_modules/mdns/lib/resolver_sequence_tasks.js:120:9)
```

it should be added:
``` + browser.on('error', err => {console.log(err)});```
to avoid server crash

Reply:
Thanks, good catch.